### PR TITLE
Simplify calendar construction for Invariant mode

### PR DESCRIPTION
**Pull request by marek-safar**
Allows illinker to save more in invariant globalization mode

Not very accurate data for browser-wasm hello world as they will change as we progress on size trimming
|Mode  | SPC Size |
|-|-|
|Default| 1449KB | 
|Invariant| 1396KB |

### CODE REVIEW

File: src/libraries/System.Private.CoreLib/src/System/Globalization/DateTimeFormatInfo.cs [hunk 1]
```diff
@@ -407,6 +407,14 @@
                     return;
                 }
 
+                if (GlobalizationMode.Invariant)
+                {
+                    if (value.ID == calendar.ID)
```
Comment (tarekgh) [inline src/libraries/System.Private.CoreLib/src/System/Globalization/DateTimeFormatInfo.cs:412]: >calendar [](start = 36, length = 8)

calendar can be null here

**Review by tarekgh — APPROVED** [review]
I added one comment you need to fix. other than that LGTM.

**Review by safern — APPROVED** [review]

File: src/libraries/System.Private.CoreLib/src/System/Globalization/CalendarData.cs [hunk 2]
```diff
@@ -324,6 +324,11 @@ private void InitializeAbbreviatedEraNames(string localeName, CalendarId calenda
 
         internal static CalendarData GetCalendarData(CalendarId calendarId)
         {
+            if (GlobalizationMode.Invariant)
+            {
+                return CalendarData.Invariant;
+            }
```
Comment (stephentoub) [inline src/libraries/System.Private.CoreLib/src/System/Globalization/CalendarData.cs:330]: Do we already return invariant data for all calendarIds, and this is just doing the check earlier/differently to help the linker trim better?

Or is this actually changing behavior when GlobalizationMode.Invariant is set?

Comment (tarekgh) [reply src/libraries/System.Private.CoreLib/src/System/Globalization/CalendarData.cs:330]: In Invariant mode, we always use the Invariant culture when requesting any culture. so, I believe we are not changing any behavior for the invariant mode here.

Comment (marek-safar) [reply src/libraries/System.Private.CoreLib/src/System/Globalization/CalendarData.cs:330]: There is no behaviour change, PR only adds few explicit checks so less code is executed under GlobalizationMode.Invariant with the same result

**Comment by jkotas** [comment]
```
    System.Globalization.Tests.InvariantModeTests.TestCultureData(cultureName: "en-US") [FAIL]
      System.ArgumentOutOfRangeException : Not a valid calendar for the given culture. (Parameter 'value')
      Actual value was System.Globalization.GregorianCalendar.
      Stack Trace:
        /_/src/libraries/System.Private.CoreLib/src/System/Globalization/DateTimeFormatInfo.cs(415,0): at System.Globalization.DateTimeFormatInfo.set_Calendar(Calendar value)
        /_/src/libraries/System.Private.CoreLib/src/System/Globalization/CultureInfo.cs(731,0): at System.Globalization.CultureInfo.get_DateTimeFormat()
        /_/src/libraries/System.Globalization/tests/Invariant/InvariantMode.cs(510,0): at System.Globalization.Tests.InvariantModeTests.TestCultureData(String cultureName)
```

**Review by safern — APPROVED** [review]
Other than one question, LGTM. @tarekgh could you take a look as well?

File: src/libraries/System.Private.CoreLib/src/System/Globalization/DateTimeFormatInfo.cs [hunk 3]
```diff
@@ -407,6 +407,17 @@ public Calendar Calendar
                     return;
                 }
 
+                if (GlobalizationMode.Invariant)
+                {
+                    if (value.ID == CalendarId.GREGORIAN)
+                    {
+                        calendar = value;
```
[the remark this answers is not in the document]
Comment (marek-safar) [reply src/libraries/System.Private.CoreLib/src/System/Globalization/DateTimeFormatInfo.cs]: Fixed.

**Review by tarekgh — APPROVED** [review]
LGTM.

File: src/libraries/System.Private.CoreLib/src/System/Globalization/CultureData.cs [hunk 4]
```diff
@@ -1671,7 +1671,7 @@ internal CalendarId[] CalendarIds
         {
             get
             {
-                if (_waCalendars == null)
+                if (_waCalendars == null && !GlobalizationMode.Invariant)
```
Comment (safern) [inline src/libraries/System.Private.CoreLib/src/System/Globalization/CultureData.cs:1674]: Is this needed? based on this https://github.com/dotnet/runtime/pull/37527/files/48393f74f45999eb53ce79c7cfda61d7c30fb691..eee299b30be940db2a9857a70aab76de6a5eb55d#diff-235473f71512930b24ea62cbe21024f6R630 `_waCalendars` shouldn't be null for `Invariant` right?

Comment (marek-safar) [reply src/libraries/System.Private.CoreLib/src/System/Globalization/CultureData.cs:1674]: Yes but illinker cannot statically detect that

Comment (safern) [reply src/libraries/System.Private.CoreLib/src/System/Globalization/CultureData.cs:1674]: Ah I see, this is just to help the linker trim this code... ok cool.

**Review by safern — APPROVED** [review]

**Pull request merged**